src/ch2.2.4/ex5/circular-list.ts

```diff
@@ -7,45 +7,37 @@ export interface ListNode<T> {
     link: ListNode<T>;
 }
 
-interface ProtoListNode<T> {
-    info: T;
-    link: ProtoListNode<T> | NullLink;
-}
-
 export enum NullLink {
     NULL,
 }
 export const Λ = NullLink.NULL;
 
-function isProtoListNode<T>(
-    n: ProtoListNode<T> | NullLink,
-): n is ProtoListNode<T> {
-    return n !== Λ;
+function nodePointingToItself<T>(info: T): ListNode<T> {
+    let node = { info, link: (Λ as any) as ListNode<T> };
```

Comment:
There's something annoying about having to commit this "white lie" to the type system. Even if it does get fixed on the very next line.

But I really can't come up with a way to do it _without_ lying to the type system. These three things together to create an impossible situation:

1. We want to type the object created as a `ListNode<T>`.
2. A `ListNode<T>` needs to have a `link` property which is a `ListNode<T>`.
3. There's no way to create an object literal whose properties get set to the object literal being created (seemingly for ordering reasons; the property values are computed first, and only then is the object created).
4.